tiledb/sm/storage_manager/context.cc

```diff
@@ -175,60 +182,80 @@ Status Context::init_thread_pools(Config* const config) {
-  uint64_t compute_concurrency_level = 0;
-  RETURN_NOT_OK(tmp_config.get<uint64_t>(
-      "sm.compute_concurrency_level", &compute_concurrency_level, &found));
-  assert(found);
-  uint64_t io_concurrency_level = 0;
-  RETURN_NOT_OK(tmp_config.get<uint64_t>(
-      "sm.io_concurrency_level", &io_concurrency_level, &found));
-  assert(found);
-
   // The "sm.num_tbb_threads" has been deprecated. Users may still be setting
   // this configuration parameter. In this scenario, we will override the
   // compute and io concurrency levels if the configured tbb threads are
   // greater.
   int num_tbb_threads = 0;
   RETURN_NOT_OK(
-      tmp_config.get<int>("sm.num_tbb_threads", &num_tbb_threads, &found));
+      config.get<int>("sm.num_tbb_threads", &num_tbb_threads, &found));
   if (found && num_tbb_threads > 0)
     max_thread_count =
         std::max(max_thread_count, static_cast<uint64_t>(num_tbb_threads));
 
-  // If 'max_thread_count' is larger than the compute or io concurrency levels,
-  // use that instead.
-  compute_concurrency_level =
-      std::max(max_thread_count, compute_concurrency_level);
-  io_concurrency_level = std::max(max_thread_count, io_concurrency_level);
-
-  // Initialize the thread pools.
-  RETURN_NOT_OK(compute_tp_.init(compute_concurrency_level));
-  RETURN_NOT_OK(io_tp_.init(io_concurrency_level));
+  max_thread_count_ret = static_cast<size_t>(max_thread_count);
 
   return Status::Ok();
 }
 
-Status Context::init_loggers(Config* const config) {
-  // If `config` is null, use a default-constructed config.
-  Config tmp_config;
-  if (config != nullptr)
-    tmp_config = *config;
+size_t Context::get_compute_thread_count(const Config& config) {
+  uint64_t max_thread_count = 0;
+  if (!get_max_thread_count(config, max_thread_count).ok()) {
+    throw std::logic_error("Cannot get max thread count");
+  }
+
+  bool found = false;
+  uint64_t compute_concurrency_level = 0;
+  if (!config
+           .get<uint64_t>(
+               "sm.compute_concurrency_level",
+               &compute_concurrency_level,
+               &found)
+           .ok()) {
+    throw std::logic_error("Cannot get compute concurrency level");
+  }
+  assert(found);
+
+  return static_cast<size_t>(
+      std::max(max_thread_count, compute_concurrency_level));
+}
 
+size_t Context::get_io_thread_count(const Config& config) {
+  uint64_t max_thread_count = 0;
+  if (!get_max_thread_count(config, max_thread_count).ok()) {
+    throw std::logic_error("Cannot get max thread count");
+  }
+
+  bool found = false;
+  uint64_t io_concurrency_level = 0;
+  if (!config
+           .get<uint64_t>(
+               "sm.io_concurrency_level", &io_concurrency_level, &found)
+           .ok()) {
+    throw std::logic_error("Cannot get max thread count");
+  }
+  assert(found);
+
+  io_concurrency_level = std::max(max_thread_count, io_concurrency_level);
+
+  return static_cast<size_t>(std::max(max_thread_count, io_concurrency_level));
```

Comment:
If `max_thread_count` means like what I expect it to mean, that should be `std::min`.

Reply:
I just recapitulated the computation as it was before in `Context::init_thread_pools()`.  The associated comment was
```c++
  // If 'max_thread_count' is larger than the compute or io concurrency levels,
  // use that instead.
```
So the current behavior definitely is to use `max_thread_count` if it is larger.  We can change that, but it would potentially change behavior.

Reply:
No need to change behavior. On the other hand, it would be good to rename `max_thread_count` to `min_thread_count` so that it says what it means.

Reply:
Though "min" is correct, I changed this to `config_thread_count` to avoid confusion of having something with `min_` in its name being computed with calls to `max_`.

Reply:
@lums658 is there a commit missing with the change in your final comment?

Reply:
Yes -- But I'm not sure what happened to it. Will re-implement and commit.